React/class-components.md

```diff
@@ -0,0 +1,110 @@
+    <div>
+      <h3>{props.fullname}</h3>
+      <p>{props.email}</p>
+    </div>
+  );
+};
+```
+
+These are called, _functional components,_ because they are composed of one function.  Functional components take in data from `props` and return JSX.  These components are easy to read, test and debug because given a set of props they always return the same JSX back.  However often we want to keep track of information over time, or the _state_ of our component.  This is why functional components are often called _Stateless Components_ in React.  They take props as input and return JSX, which can be a mixture of nested components and HTML.
+
+If you remember back to Ruby, we built objects which combined data, which could change over time, with functionality by defining classes.  We can build React components which can do the same!
+
+## Class Components 
+
+We have already seen a class component `create-react-app` generates one for us called `App`.  
+
+```javascript
+class App extends Component {
+  render() {
+    return (
+      <div className="App">
+       ...
+      </div>
+    );
+  }
+}
+```
+
+This class extends `Component` a class defined in the React library and inherits a bunch of functionality we will go into further in following lectures.  All classes **must** have a `render` method which returns a block of JSX like the functional components we have created thus far.  
+
+### What about `props`
+
+Because a class component is not a single function, any props must be passed into the constructor.
+
+We can see an example by re-writing our Student component as a class.
+
+```javascript
+// components/Student.js
+import React from 'react';
+
+class Student extends React.Component {
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    return (
+      <div>
+        <h3>{this.props.fullname}</h3>
+        <p>{this.props.email}</p>
+      </div>
+    );
+  }  
+}
+
+export default Student;
+```
+
+In the Student class, `props` are passed into the contructor as arguments and saved as an instance variable by the superclass, Component.  In a class component you can always access props from any method with `this.props`, just like you could access a ruby instance variable with the `@` character.  Because this is the default behavior of a React component, you can leave off the constructor, if this is all you are doing.
+
+**Questions:**
+
+1. What is the superclass for `Student`?
+1. What is accomplished with `super(props)`?
+
+### Is it `extends React.Component` or `extends Component`
+
+You may have noticed that `App` inherits from `Component` while the `Student` inherits from `React.Component`.  They are both inheriting from the same class, but in `App.js` we have _destructured_ the `React` object with `import React, { Component } from 'react';`.  The `{ Component }` part of the line is the equivalent to:  `const Component = React.Component`.  It's a way of assigning elements of an object to individual variables.  
+
```

Comment:
I like this explanation!

Reply:
I also think this is a good explanation of destructuring within an `import` statement (I think technically it's separate from the destructuring used in other parts of the JS syntax).

However, I'm not sure this is a great place to explain this concept. It's embedded like 2/3rds of the way through this lecture, and the only connection it has to the rest of the lecture is that we've chosen to use the non-destructuring syntax in our code snippets here.

If we instead agreed to always use the `import { React, Component} from 'react';` version instead, then we wouldn't have any discrepancy in how we write the parent class, and we could leave this explanation out entirely.